Quality Configuration

Dhruv-0-Arora opened the pull request:
## Description
Creates a new panel that allows users to have more control over their graphics settings (specifically cascading shadows). Allows for enhancing graphics or improving performance.

<img width="376" alt="Screenshot 2025-06-17 at 4 53 22 PM" src="https://github.com/user-attachments/assets/c633d272-67bc-4562-94da-a0ac9c6561b9" />

<img width="345" alt="Screenshot 2025-06-17 at 4 56 36 PM" src="https://github.com/user-attachments/assets/26f9a7e7-9323-45ce-915a-f0a49687245d" />

### Features
- [x] Change light intensity
- [x] Toggle cascading shadows
- [x] Change Max Far (zoom out distance before the shadows are unrendered)
- [x] Change Shadow Map Size (shadow texture size/detail)
- [x] Reset cascading shadows values button
- [x] Anti-aliasing toggle (will hot reload the page)

### Notes
- Removed previous quality dropdown
<img width="346" alt="Screenshot 2025-06-17 at 4 56 05 PM" src="https://github.com/user-attachments/assets/603144fe-95b5-4dc0-a420-338143088f38" />

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1780)

Dhruv-0-Arora committed:
Added basic UI

Dhruv-0-Arora committed:
Power Preference Configuration

Dhruv-0-Arora committed:
Changing Anti-aliasing implementation however it doesn't work

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1780/quality-panel

Dhruv-0-Arora committed:
Working antialias switch switch

Dhruv-0-Arora committed:
Fixed preference system implementation and created working anti-aliasing customization

Dhruv-0-Arora committed:
Added little info bubble

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1780/quality-panel

Dhruv-0-Arora committed:
Added unsaved deletion

LucaHaverty commented:
Hovering over the '?' icon causes a jittering behavior for me because it shifts the panel over slightly

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1780/quality-panel

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1780/quality-panel

Dhruv-0-Arora committed:
Updated the UI and functionality. Need to iron out bugs

LucaHaverty commented:
Is this PR waiting for something else to be merged in or on hold?

LucaHaverty reviewed (commented):
I feel like the vast majority of users won't know what things like "cascades" and "shadow map size" are, so in my opinion it would be better to have preset quality levels that we know are stable rather than give them control over all the specific settings with the potential of breaking something or ending up with crazy artifacts.

BrandonPacewic commented:
> I feel like the vast majority of users won't know what things like "cascades" and "shadow map size" are, so in my opinion it would be better to have preset quality levels that we know are stable rather than give them control over all the specific settings with the potential of breaking something or ending up with crazy artifacts.

I agree, but I do think we should eventually have both as an option. Similar to how most games give you low, medium, and high presets but still let you change whatever specific option you want. (At your own risk of course).

BrandonPacewic reviewed (commented):
> I feel like the vast majority of users won't know what things like "cascades" and "shadow map size" are, so in my opinion it would be better to have preset quality levels that we know are stable rather than give them control over all the specific settings with the potential of breaking something or ending up with crazy artifacts.

I agree, but I do think we should eventually have both as an option. Similar to how most games give you low, medium, and high presets but still let you change whatever specific option you want. (At your own risk of course).

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1780/quality-panel

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1780/quality-panel

Dhruv-0-Arora committed:
Fixing build error

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1780/quality-panel

Dhruv-0-Arora committed:
Fixed build errors and removed quality dropdown
Additionally ran formatting. Next steps are to fix concurrent changes to cascading shadows

Dhruv-0-Arora committed:
Finishing the quality controls functionality and fixing the injection update

Dhruv-0-Arora committed:
fix: changing the maximum value for the "max far" variable

BrandonPacewic commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -230,42 +233,89 @@ class SceneRenderer extends WorldSystem {
+        this.CreateCSM(settings)
+        this.SetupCSMMaterials()
+
+        // this._scene.traverse(object => {
+        //     if (object.material) {
+        //         if (Array.isArray(object.material)) {
+        //             object.material.forEach(mat => {
+        //                 mat.needsUpdate = true
+        //             })
+        //         } else {
+        //             object.material.needsUpdate = true
+        //         }
+        //     }
+        // })
```
Does this update still not need to be done?

Dhruv-0-Arora replied:
Removed the comments

BrandonPacewic commented on fission/.eslintrc.cjs:
If possible we should avoid formatting these files beyond what has already been done. We don't enforce style here so your essentially putting your own auto formatting standards onto the repo.

Dhruv-0-Arora replied:
Ill revert the changes to .eslintrc.cjs and ill keep the changes for the readme and postcss config

BrandonPacewic commented on fission/README.md:
Same as above. I'm not necessarily opposed to letting these changes go through since this file was already poorly formatted.

BrandonPacewic commented on fission/postcss.config.js:
Same as before. We should leave this change in though since we use double quotes and trailing commas anyways.

BrandonPacewic commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -86,11 +85,11 @@ class SceneRenderer extends WorldSystem {
             stencil: false,
-            depth: false,
+            depth: !PreferencesSystem.getGraphicsPreferences().antiAliasing,
+            // logarithmicDepthBuffer: true,
```
This still need to be here? Should add a comment if we want it to be a reference for later.

BrandonPacewic commented on fission/src/ui/panels/GraphicsSettingsPanel.tsx:
```diff
@@ -0,0 +1,184 @@
+                    <></>
+                )}
+                <div className="flex items-center justify-center mt-1 mb-0.5 mx-[5%]">
+                    <SectionLabel size={LabelSize.Medium} className="text-center">
+                        Hot-reload Settings
+                    </SectionLabel>
```
I'm partial to the opinion that this label should be removed. At first I thought it was some sort of button or toggle and it seemed out of place.

BrandonPacewic commented on fission/src/ui/panels/GraphicsSettingsPanel.tsx:
```diff
@@ -0,0 +1,184 @@
+                    </SectionLabel>
+                </div>
+                <SectionDivider />
+                <Checkbox
+                    label="Anti-Aliasing"
+                    defaultState={antiAliasing}
+                    onClick={checked => {
+                        // saving the new preference
+                        setAntiAliasing(checked)
+                        setReload(true)
+                    }}
+                    tooltipText="Requires a browser refresh"
+                />
```
Should probably have some sort of user confirmation here before you auto refresh and reset the current scene. Could do this by adding some sort of confirmation after the user clicks accept with a change to this option or could be saved with some sort of "Refresh is required to apply some settings" message.

BrandonPacewic commented on fission/tailwind.config.mjs:
Same as other formatting changes comment. While I agree that this file is also poorly formatted this should probably be its own pr.

Dhruv-0-Arora replied:
Reverted

BrandonPacewic commented on fission/vite.config.ts:
Same as other formatting changes comment.

Dhruv-0-Arora replied:
Reverted

BrandonPacewic reviewed (changes requested):
Additionally I don't know if I'm the biggest fan of having the panel for settings open in a different place compared to the main settings panel. It does make sense from a visual perspective so you can see what changes your actually making. Open to feedback on that.

Dhruv-0-Arora committed:
Added requested changes and reverted formatting changes
chore: removed unused commented code
moved panel location to the middle of the screen

BrandonPacewic reviewed (changes requested):
Small merge conflict.

Dhruv-0-Arora committed:
Merging dev

PepperLola commented on fission/src/systems/scene/SceneRenderer.ts:
```diff
@@ -90,20 +89,19 @@ class SceneRenderer extends WorldSystem {
 
-        // Adding the lighting using quality preferences
-        this.ChangeLighting(PreferencesSystem.getGlobalPreference<string>("QualitySettings"))
+        // Adding the lighting uisng quality preferences
```
```suggestion
        // Adding the lighting using quality preferences
```

PepperLola reviewed (commented):
I think it looks good. We might want to eventually make the settings tabbed rather than opening new panels with buttons but that should happen after the UI refactor. I suggested a change for a typo but other than that I think it's ready to merge.

Do we want to limit the maximum values that someone can choose? I think the upper end of the texture size values is almost impossible to run at an acceptable frame rate (especially since most people will probably be using chromebooks), so would it make sense to limit options that would tank performance like that to ones that are actually feasible to run? Especially if the settings are persisted, because it would be hard to reset the values if it takes a minute to click each button.

Dhruv-0-Arora committed:
Merging and fixing spelling error

Dhruv-0-Arora commented:
> I think it looks good. We might want to eventually make the settings tabbed rather than opening new panels with buttons but that should happen after the UI refactor. I suggested a change for a typo but other than that I think it's ready to merge.
> 
> Do we want to limit the maximum values that someone can choose? I think the upper end of the texture size values is almost impossible to run at an acceptable frame rate (especially since most people will probably be using chromebooks), so would it make sense to limit options that would tank performance like that to ones that are actually feasible to run? Especially if the settings are persisted, because it would be hard to reset the values if it takes a minute to click each button.

@PepperLola 
Currently I have the max shadow map size set to the renderers maxTextureSize capability, so on slower computers it will automatically limit the user. We could also manually set the value.

![Screenshot 2025-06-19 at 3 39 01 PM](https://github.com/user-attachments/assets/a106f2be-e056-4ff3-b9d0-57aa36d8d1a5)

Dhruv-0-Arora committed:
Merge dev conflict

BrandonPacewic reviewed (approved):
I think it's time to move onto something else. I have a few remaining issues that I think we should address after the UI refactor:

1. I agree that this should probably be a tab within the setting panel itself, rather than opening a new panel.
2. I think that we should consider splitting up the settings into collections (maybe a drop down) of basic and advanced.
3. I'm not that big of a fan of how we currently handle settings that need a refresh to take effect, again something that will be easier to handle (or at least implement) post UI refactor.
4. I think removing the presets all together is not what we want the final version of this to look like. I'm adding this as an additional task blocked by AARD-1781.

Since all my problems are essentially just UI nitpicks and the functionality is here I think its appropriate to make these changes after we finish the UI refactor.

Dhruv-0-Arora committed:
Merging Dev

PepperLola reviewed (approved):
I think it looks good except for the max texture size issues. In my experience it isn't properly limited and setting it too high can cause the app to become entirely unusable. I'm willing to leave that until later though, but I do think it should be addressed at some point.

BrandonPacewic commented:
> I think it looks good except for the max texture size issues. In my experience it isn't properly limited and setting it too high can cause the app to become entirely unusable. I'm willing to leave that until later though, but I do think it should be addressed at some point.

Noted AARD-1928.
